security/advancedtls/pemfile_provider_test.go

```diff
@@ -0,0 +1,220 @@
+	cs := &certStore{}
+	err := cs.loadCerts()
+	if err != nil {
+		t.Fatalf("cs.loadCerts() failed: %v", err)
+	}
+	tests := []struct {
+		desc         string
+		options      *PEMFileProviderOptions
+		wantKmStage0 certprovider.KeyMaterial
+		wantKmStage1 certprovider.KeyMaterial
+		wantKmStage2 certprovider.KeyMaterial
+	}{
+		{
+			desc: "use identity certs and root certs",
+			options: &PEMFileProviderOptions{
+				CertFile:  "not_empty_cert_file",
+				KeyFile:   "not_empty_key_file",
+				TrustFile: "not_empty_trust_file",
+			},
+			wantKmStage0: certprovider.KeyMaterial{Certs: []tls.Certificate{cs.clientPeer1}, Roots: cs.serverTrust1},
+			wantKmStage1: certprovider.KeyMaterial{Certs: []tls.Certificate{cs.clientPeer1}, Roots: cs.serverTrust1},
+			wantKmStage2: certprovider.KeyMaterial{Certs: []tls.Certificate{cs.clientPeer2}, Roots: cs.serverTrust2},
+		},
+		{
+			desc: "use identity certs only",
+			options: &PEMFileProviderOptions{
+				CertFile: "not_empty_cert_file",
+				KeyFile:  "not_empty_key_file",
+			},
+			wantKmStage0: certprovider.KeyMaterial{Certs: []tls.Certificate{cs.clientPeer1}},
+			wantKmStage1: certprovider.KeyMaterial{Certs: []tls.Certificate{cs.clientPeer1}},
+			wantKmStage2: certprovider.KeyMaterial{Certs: []tls.Certificate{cs.clientPeer2}},
+		},
+		{
+			desc: "use trust certs only",
+			options: &PEMFileProviderOptions{
+				TrustFile: "not_empty_trust_file",
+			},
+			wantKmStage0: certprovider.KeyMaterial{Roots: cs.serverTrust1},
+			wantKmStage1: certprovider.KeyMaterial{Roots: cs.serverTrust1},
+			wantKmStage2: certprovider.KeyMaterial{Roots: cs.serverTrust2},
+		},
+	}
+	for _, test := range tests {
+		test := test
+		testInterval := 200 * time.Millisecond
+		test.options.IdentityInterval = &testInterval
+		test.options.RootInterval = &testInterval
+		t.Run(test.desc, func(t *testing.T) {
+			stage := &stageInfo{}
+			readKeyCertPairFunc = func(certFile, keyFile string) (tls.Certificate, error) {
+				switch stage.read() {
+				case 0:
+					return cs.clientPeer1, nil
+				case 1:
+					return tls.Certificate{}, fmt.Errorf("error occurred while reloading")
+				case 2:
+					return cs.clientPeer2, nil
+				default:
+					return tls.Certificate{}, fmt.Errorf("test stage not supported")
+				}
+			}
+			readTrustCertFunc = func(trustFile string) (*x509.CertPool, error) {
+				switch stage.read() {
+				case 0:
+					return cs.serverTrust1, nil
+				case 1:
+					return nil, fmt.Errorf("error occurred while reloading")
+				case 2:
+					return cs.serverTrust2, nil
+				default:
+					return nil, fmt.Errorf("test stage not supported")
+				}
+			}
+			defer restoreMockFunctionsBack()
+			provider, err := NewPEMFileProvider(test.options)
+			if err != nil {
+				t.Fatalf("NewPEMFileProvider failed: %v", err)
+			}
+			defer provider.Close()
+			//// ------------------------Stage 0------------------------------------
+			time.Sleep(1 * time.Second)
+			gotKM, err := provider.KeyMaterial(context.Background())
```

Comment:
Please create a context with a valid small deadline, so that even if something goes wrong the test will fail in a reasonable amount of time, instead of hanging forever.